src/Features/VisualBasic/Portable/ConvertLinq/VisualBasicConvertLinqQueryToLinqMethodProvider.vb

```diff
@@ -0,0 +1,176 @@
+                Dim lambdaHeader = SyntaxFactory.LambdaHeader(SyntaxKind.FunctionLambdaHeader, Nothing, Nothing, SyntaxFactory.Token(SyntaxKind.FunctionKeyword), SyntaxFactory.ParameterList().AddParameters(parameter), asClause:=Nothing)
+                Dim lambda = SyntaxFactory.SingleLineLambdaExpression(SyntaxKind.SingleLineFunctionLambdaExpression, lambdaHeader, bodyExpression.WithoutTrailingTrivia())
+                Dim argument = SyntaxFactory.SimpleArgument(lambda)
+                Dim arguments = New SeparatedSyntaxList(Of ArgumentSyntax)().Add(argument)
+                Dim argumentList = SyntaxFactory.ArgumentList(arguments)
+                Return CreateInvocationExpression(expression, keyword, argumentList)
+            End Function
+
+            Private Function ProcessPartitionClause(expression As ExpressionSyntax, bodyExpression As ExpressionSyntax, keyword As String) As ExpressionSyntax
+                Dim argument = SyntaxFactory.SimpleArgument(bodyExpression.WithoutTrailingTrivia())
+                Dim arguments = New SeparatedSyntaxList(Of ArgumentSyntax)().Add(argument)
+                Dim argumentList = SyntaxFactory.ArgumentList(arguments)
+
+                Return CreateInvocationExpression(expression, keyword, argumentList)
+            End Function
+
+            Private Function CreateInvocationExpression(expression As ExpressionSyntax, keyword As String, Optional argumentList As ArgumentListSyntax = Nothing) As ExpressionSyntax
+                Dim keywordIdentifier = SyntaxFactory.IdentifierName(keyword)
+                Dim memberAccessExpression = SyntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, expression, SyntaxFactory.Token(SyntaxKind.DotToken), keywordIdentifier)
+                Return SyntaxFactory.InvocationExpression(memberAccessExpression, argumentList)
+            End Function
+
+            Private Function ProcessOrderByClause(expression As ExpressionSyntax, identifier As SyntaxToken, orderByClause As OrderByClauseSyntax) As ExpressionSyntax
+                Dim isFirst = True
+                For Each ordering In orderByClause.Orderings
+                    Dim keyword As String
+                    Select Case (ordering.Kind())
+                        Case SyntaxKind.AscendingOrdering : keyword = If(isFirst, "OrderBy", "ThenBy")
+                        Case SyntaxKind.DescendingOrdering : keyword = If(isFirst, "OrderByDescending", "ThenByDescending")
+                        Case Else : Return Nothing
+                    End Select
+                    expression = CreateInvocationExpression(expression, identifier, ordering.Expression, keyword)
+                    isFirst = False
+                Next
+
+                Return expression
+            End Function
+        End Class
+
+        ' VB QueryExpressionSyntax consists of a collection of QueryClause. Each of them can change the context for all descendants. 
+        Private Class ConversionContext
```

Comment:
why do we need this type if it just wraps a token?

Reply:
With more cases to implement, this class will be extended to [https://github.com/dotnet/roslyn/blob/d4dab355b96955aca5b4b0ebf6282575fad78ba8/src/Compilers/CSharp/Portable/Binder/Binder.QueryTranslationState.cs](https://github.com/dotnet/roslyn/blob/d4dab355b96955aca5b4b0ebf6282575fad78ba8/src/Compilers/CSharp/Portable/Binder/Binder.QueryTranslationState.cs)

---
In reply to: [167387986](https://github.com/dotnet/roslyn/pull/24759#discussion_r167387986) [](ancestors = 167387986)